File: AIML/officer_view.py
"""Aggregate existing SAR and yield services for the Officer View."""

from __future__ import annotations

import datetime as dt
import sys
from pathlib import Path
from typing import Any, Dict, Iterable, List


BASE_DIR = Path(__file__).resolve().parent
SAR_DIR = BASE_DIR.parent / "sar_processing"
if str(SAR_DIR) not in sys.path:
    sys.path.insert(0, str(SAR_DIR))

from anomaly_detection import detect_flood, detect_moisture_anomaly
from config import MOISTURE_DELTA_THRESHOLD
from feature_extractor import compute_deltas
from gee_client import get_sar_means
from yield_service.app import YieldPredictionRequest, predict as predict_yield


ACRE_TO_HECTARE = 0.4047
MODEL_YEAR = "2020-21"


def _season_for(date: dt.date) -> str:
    """Use the same season names accepted by the existing yield endpoint."""
    if 6 <= date.month <= 10:
        return "Kharif"
    if date.month >= 11 or date.month <= 3:
        return "Rabi"
    return "Summer"


def _point_geometry(farmer: Dict[str, Any]) -> Dict[str, Any]:
    """Adapt the seed profile's GPS coordinate to the SAR client's GeoJSON input."""
    return {
        "type": "Point",
        "coordinates": [float(farmer["longitude"]), float(farmer["latitude"])],
    }


def _area_in_hectares(farmer: Dict[str, Any]) -> float:
    area = float(farmer.get("land_area") or 0)
    if area <= 0:
        raise ValueError(f"Farmer {farmer.get('id', 'unknown')} has no valid land_area")
    return area * ACRE_TO_HECTARE if str(farmer.get("land_unit", "")).lower() == "acre" else area


def _risk_score(flood_flag: bool, moisture_anomaly: str) -> int:
    """Map existing SAR classifications to a compact Officer View display score."""
    if flood_flag:
        return 100
    return {"high": 70, "low": 55, "normal": 15}.get(moisture_anomaly, 15)


def _alert_urgency(flood_flag: bool, moisture_anomaly: str) -> str:
    """Express the current SAR result using the dashboard's existing urgency labels."""
    if flood_flag:
        return "critical"
    if moisture_anomaly == "high":
        return "high"
    if moisture_anomaly == "low":
        return "medium"
    return "low"


def _bounded_score(value: Any) -> float:
    """Keep a supplied 0-100 signal safe for the Officer View formula."""
    try:
        return max(0.0, min(100.0, float(value)))
    except (TypeError, ValueError):
        return 0.0


def computeUrgencyScore(farmer: Dict[str, Any]) -> float:
    """Return the Officer View urgency score for one A3-enriched farmer.

    A3 exposes its SAR signal as ``riskScore``. It does not yet expose a
    disease-pressure score or a pending-insurance-claim flag, so those terms
    safely contribute zero until their source data is connected.
    """
    flood_risk = _bounded_score(farmer.get("riskScore"))
    disease_pressure = 0.0
    pending_insurance_claim = False
    score = (
        flood_risk * 0.4
        + disease_pressure * 0.3
        + (30 if pending_insurance_claim else 0) * 0.3
    )
    return round(max(0.0, min(100.0, score)), 2)


def getOfficerViewData(farmers: Iterable[Dict[str, Any]]) -> Dict[str, Any]:
    """Enrich farmer profiles with existing SAR and yield-model outputs.

    ``projectedYield`` is total expected production in tonnes, so the aggregate
    can be summed across farms. SAR is sampled for the current and preceding
    seven-day windows at each farmer's registered GPS point.
    """
    today = dt.date.today()
    seven_days = dt.timedelta(days=7)
    enriched_farmers: List[Dict[str, Any]] = []

    for farmer in farmers:
        geometry = _point_geometry(farmer)
        current_sar = get_sar_means(geometry, today - seven_days, today)
        previous_sar = get_sar_means(geometry, today - (seven_days * 2), today - seven_days)
        deltas = compute_deltas(current_sar, previous_sar)
        moisture_anomaly = detect_moisture_anomaly(
            deltas["vv_delta_7d"], MOISTURE_DELTA_THRESHOLD
        )
        flood_flag = detect_flood(current_sar.get("vv_mean"), current_sar.get("vh_mean"))

        area_hectares = _area_in_hectares(farmer)
        yield_result = predict_yield(
            YieldPredictionRequest(
                state=str(farmer["state"]),
                district=str(farmer["district"]),
                crop=str(farmer["primary_crop"]),
                year=MODEL_YEAR,
                season=_season_for(today),
                area=area_hectares,
            )
        )
        projected_yield = float(yield_result["predicted_yield"]) * area_hectares

        enriched_farmers.append(
            {
                **farmer,
                "riskScore": _risk_score(flood_flag, moisture_anomaly),
                "projectedYield": round(projected_yield, 2),
                "alertUrgency": _alert_urgency(flood_flag, moisture_anomaly),
            }
        )

    ranked_farmers = [
        {**farmer, "urgencyScore": computeUrgencyScore(farmer)}
        for farmer in enriched_farmers
    ]
    ranked_farmers.sort(key=lambda farmer: farmer["urgencyScore"], reverse=True)

    aggregate = {
        "totalProjectedYield": round(
            sum(farmer["projectedYield"] for farmer in ranked_farmers), 2
        ),
        "farmersAtCriticalRisk": sum(
            farmer["alertUrgency"] == "critical" for farmer in ranked_farmers
        ),
    }
    return {
        "farmers": ranked_farmers,
        "priorityQueue": ranked_farmers[:3],
        "aggregate": aggregate,
    }

File: client/src/lib/officerView.js
/**
 * client/src/lib/officerView.js
 * JavaScript mirror of AIML/officer_view.py:
 * Provides getOfficerViewData() and computeUrgencyScore()
 * over seed-farmers.json with deterministic ranking.
 */

import seedFarmers from '../seed-farmers.json'

const ACRE_TO_HECTARE = 0.4047

/** Base yield by crop in tonnes per hectare (mirrors yield_service outputs for Gujarat) */
const BASE_CROP_YIELDS = {
  Groundnut: 2.65,
  Maize: 3.85,
  Soyabean: 2.15,
  Rice: 4.10,
  Wheat: 3.60,
  Onion: 18.20,
  Potato: 22.50,
}

function boundedScore(val) {
  const n = Number(val)
  if (isNaN(n)) return 0
  return Math.max(0, Math.min(100, n))
}

export function areaInHectares(farmer) {
  const area = Number(farmer?.land_area || 0)
  if (area <= 0) return 0
  const unit = String(farmer?.land_unit || '').toLowerCase()
  return unit === 'acre' ? area * ACRE_TO_HECTARE : area
}

export function riskScore(floodFlag, moistureAnomaly) {
  if (floodFlag) return 100
  if (moistureAnomaly === 'high') return 70
  if (moistureAnomaly === 'low') return 55
  return 15
}

export function alertUrgency(floodFlag, moistureAnomaly) {
  if (floodFlag) return 'critical'
  if (moistureAnomaly === 'high') return 'high'
  if (moistureAnomaly === 'low') return 'medium'
  return 'low'
}

/**
 * Computes urgency score (0-100) exactly matching AIML/officer_view.py:
 * flood_risk (40%) + disease_pressure (30%) + pending_claim (30 pts * 0.3)
 */
export function computeUrgencyScore(farmer) {
  const floodRisk = boundedScore(
    farmer.riskScore !== undefined
      ? farmer.riskScore
      : farmer.flood_flag
        ? 100
        : farmer.moisture_anomaly === 'high'
          ? 70
          : 15
  )

  const diseasePressure = boundedScore(
    farmer.disease_pressure !== undefined
      ? farmer.disease_pressure
      : String(farmer.disease_risk).toLowerCase() === 'high'
        ? 85
        : String(farmer.disease_risk).toLowerCase() === 'moderate'
          ? 45
          : 0
  )

  const pendingInsuranceClaim = Boolean(farmer.pending_insurance_claim || farmer.pending_claim)

  const score = (floodRisk * 0.4) + (diseasePressure * 0.3) + ((pendingInsuranceClaim ? 30 : 0) * 0.3)
  return Math.round(Math.max(0, Math.min(100, score)) * 100) / 100
}

/**
 * Human-readable reason for why the farmer was placed in priority queue
 */
export function urgencyReason(farmer) {
  const parts = []
  if (farmer.flood_flag || farmer.riskScore >= 80) parts.push('Active flood risk')
  if (farmer.pending_insurance_claim || farmer.pending_claim) parts.push('Pending insurance claim')
  if (String(farmer.disease_risk).toLowerCase() === 'high' || farmer.disease_pressure >= 70) {
    parts.push('High disease pressure')
  } else if (String(farmer.disease_risk).toLowerCase() === 'moderate' || farmer.disease_pressure >= 40) {
    parts.push('Moderate disease pressure')
  }
  if (farmer.moisture_anomaly === 'high') parts.push('High SAR moisture anomaly')
  if (farmer.alert_count && farmer.alert_count > 0) parts.push(`${farmer.alert_count} active alerts`)

  return parts.length ? parts.join(' + ') : 'Routine monitoring (baseline SAR)'
}

/**
 * Enriches farmer list with SAR classifications, yield estimates, and urgency scores.
 * Exactly mirrors getOfficerViewData in AIML/officer_view.py.
 */
export function getOfficerViewData(farmers = seedFarmers) {
  const rawList = Array.isArray(farmers) ? farmers : seedFarmers

  const enriched = rawList.map(farmer => {
    const floodFlag = Boolean(farmer.flood_flag)
    const moistureAnomaly = farmer.moisture_anomaly || 'normal'

    const areaHa = areaInHectares(farmer)
    const cropYieldPerHa = BASE_CROP_YIELDS[farmer.primary_crop] || 2.80
    const projectedYield = Math.round((cropYieldPerHa * areaHa) * 100) / 100

    const rScore = farmer.riskScore !== undefined ? Number(farmer.riskScore) : riskScore(floodFlag, moistureAnomaly)
    const urgency = farmer.alertUrgency || alertUrgency(floodFlag, moistureAnomaly)

    const baseObj = {
      ...farmer,
      area_hectares: Math.round(areaHa * 100) / 100,
      flood_flag: floodFlag,
      moisture_anomaly: moistureAnomaly,
      riskScore: rScore,
      projectedYield,
      alertUrgency: urgency,
    }

    const uScore = computeUrgencyScore(baseObj)
    return {
      ...baseObj,
      urgencyScore: uScore,
      reason: urgencyReason(baseObj),
      alert_count: farmer.alert_count || (uScore >= 70 ? 3 : uScore >= 40 ? 1 : 0),
    }
  })

  // Sort descending by urgencyScore
  const rankedFarmers = [...enriched].sort((a, b) => b.urgencyScore - a.urgencyScore)

  const aggregate = {
    totalProjectedYield: Math.round(rankedFarmers.reduce((sum, f) => sum + f.projectedYield, 0) * 100) / 100,
    farmersAtCriticalRisk: rankedFarmers.filter(f => f.alertUrgency === 'critical' || f.urgencyScore >= 70 || f.riskScore >= 80).length,
  }

  return {
    farmers: rankedFarmers,
    priorityQueue: rankedFarmers.slice(0, 3),
    aggregate,
  }
}
